### PR TITLE
Stabilize tests and tooling: hermetic deps, deterministic timing, safer checkpoint writes

### PR DESCRIPTION
**Pull request by RC219805**
Switched elapsed-time measurement to a monotonic clock to eliminate timing flake.
Prevented checkpoint temp-file collisions by writing to unique temp names before rename.
Anchored git metadata capture to the provided path to avoid “leaking” repo info from CWD.
Added safe MPS autocast fallback when unsupported at runtime.
Hardened tests for hermetic/offline environments:
Gate optional deps (transformers / DA3) with skips when unavailable.
Skip VGG19-weight-dependent training tests when weights aren’t cached offline.
Force CPU for backend-fragile MaterialsV3 E2E/edge-case tests for determinism.
Verification
pytest: 2533 passed, 219 skipped
ruff check src: clean (ran with --no-cache due to cache write restrictions)

### CODE REVIEW

File: tests/test_training_infrastructure.py [hunk 1]
```diff
@@ -299,6 +315,12 @@ class TestModelCheckpoints:
 
     def test_checkpoint_saving(self):
         """Test checkpoint can be saved"""
+        from torch.hub import get_dir
+
+        weights_path = Path(get_dir()) / "checkpoints" / "vgg19-dcbb9e9d.pth"
+        if not weights_path.exists():
+            pytest.skip("VGG19 weights not available offline")
```
[the remark this answers is not in the document]
Comment (RC219805) [reply tests/test_training_infrastructure.py]: Acknowledged. The VGG19 offline-weights guard is duplicated in TestModelCheckpoints. Given feature-freeze constraints, I’m keeping this PR minimal and leaving the duplication as-is (behavior is correct and CI is green). I’ll follow up post-freeze with a small refactor to reuse the _skip_if_no_vgg_weights() helper pattern already used in TestDepthNormalsIntegration / TestLPIPSIntegration.
